Added typescript type definitions for typescript and fixes Model.distinct() (close #52)

xrip opened the pull request:
Those little fixes would be very useful for community.
I've using chachegoose with typescript for a long time and want to share my definitions for community. Also, here is some fix for .distinct operation which not needs an hydrate it works like .lean

xrip committed:
#fix: Model.distinct('field') now works properly

xrip committed:
+add: typescript type definitions added

xrip committed:
-removed: unnecessary build

xrip committed:
Update README.md

xrip committed:
Update README.md

xrip committed:
-removed: unnecessary build

xrip committed:
Merge remote-tracking branch 'origin/master'

xrip committed:
-removed: unnecessary build

riderx commented:
Need it :)

Akhil-bridge reviewed (commented):
@xrip It showing me this error after taken your pull request
TS2551: Property 'cache' does not exist on type 'DocumentQuery<IUser, IUser, {}>'. Did you mean 'catch'
![image](https://user-images.githubusercontent.com/34673229/79839340-021c4780-83d2-11ea-8253-24c915cfb0b4.png)

And see this error
![image](https://user-images.githubusercontent.com/34673229/79843973-a5705b00-83d8-11ea-93a5-4f095004f042.png)

xrip commented:
@Akhil-bridge Did you extends global mongoose object with cache goose? Like 
```ts
import mongoose from 'mongoose';
import cachegoose from 'cachegoose';
import redis from '../redis';

const mongo = function mongo(config: string): mongoose.Connection {
    const options = {
        useUnifiedTopology: true,
        useNewUrlParser: true,
        useCreateIndex: true,
        useFindAndModify: false,
    };

    mongoose.connect(config, options);

    cachegoose(mongoose, redis);

    return mongoose.connection;
};

export default mongo(process.env.MONGODB as string);
```

Akhil-bridge commented:
Thanks xrip
This worked..!!
`import * as cachegoose from 'cachegoose';`
Changed according to the below solution
https://stackoverflow.com/questions/56238356/understanding-esmoduleinterop-in-tsconfig-file

xrip committed:
Fixes orFail type definitions issue, based on
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34205#issuecomment-621976826

xrip committed:
Fixes orFail type definitions issue, based on
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34205#issuecomment-621976826

xrip committed:
Fix same models with different db cache overlap.

xrip committed:
Fix same models with different db cache overlap.

xrip committed:
Fix same models with different db cache overlap.

Akhil-bridge commented:
Could you please merge this @boblauer ?

Akhil-bridge commented:
@xrip do you know how to keep node_modules updated with the changes in the forked repo? I don't think these changes been merged soon.

xrip commented:
You can add package directly from my github repo e.g 
'''
npm install https://github.com/xrip/cachegoose

xrip committed:
Ability to provide credentials for cache engine or cache engine client itself.

devcshort commented:
@xrip I seem to get a bunch of build errors when trying to use your repo.

```
node_modules/cachegoose/index.d.ts:21:2 - error TS2666: Exports and export assignments are not permitted in module augmentations.

21  export = cachegoose;
    ~~~~~~

node_modules/cachegoose/index.d.ts:29:9 - error TS2371: A parameter initializer is only allowed in a function or constructor implementation. 

29      cache(ttl: number = 60, customKey: string = ''): this
              ~~~~~~~~~~~~~~~~

node_modules/cachegoose/index.d.ts:29:27 - error TS2371: A parameter initializer is only allowed in a function or constructor implementation.

29      cache(ttl: number = 60, customKey: string = ''): this
                                ~~~~~~~~~~~~~~~~~~~~~~

node_modules/cachegoose/index.d.ts:30:9 - error TS2371: A parameter initializer is only allowed in a function or constructor implementation. 

30      cache(customKey: string = ''): this
              ~~~~~~~~~~~~~~~~~~~~~~

node_modules/cachegoose/index.d.ts:31:9 - error TS2371: A parameter initializer is only allowed in a function or constructor implementation.

31      cache(ttl: number = 60): this
```

xrip committed:
+ github npm update

xrip committed:
Update README.md
